### PR TITLE
Secure the cookie: Passing the erlang cookie per environment.

### PR DESCRIPTION
**Pull request by TypedLambda**
Passing the Erlang cookie in process arguments leaks the cookie in process listings.
as arguments can be observed by process monitoring tools.

Passing secrets via environment variables does not have this problem.

This patch adds support for configuring the use of an **ERLANG_COOKIE** environment variable. Passing via environment is made configurable for CNodes.

```elixir
        # config.exs
        config :bundlex :cookie_env  true # global setting false by default
        config :my_app  :cookie_env  true # override per application
        config :my_app  :cookie_env  %{a_native_name: true} # individualy
```
when set to true, the CNode is started with an empty cookie argument, and the ERLANG_COOKIE variable set.

* does not change the default behavior
* updated the the example CNode to detect an empty cookie argument and use the environment variable in this case.
* Added test cases for the updated behavior

### CODE REVIEW

File: test_projects/example/test/example_test.exs
Comment (mickel8) [inline test_projects/example/test/example_test.exs:31]:
```suggestion
  test ":cookie_env mechanism"
```

File: lib/bundlex/cnode.ex
Comment (mickel8) [inline lib/bundlex/cnode.ex:58]: I would prefix it with BUNDLEX so that we have `BUNDLEX_ERLANG_COOKIE`.

File: lib/bundlex/cnode/server.ex
Comment (mickel8) [inline lib/bundlex/cnode/server.ex:15]:
```suggestion
    cookie = unless opts.env?, do: Node.get_cookie(), else: ""
```

File: lib/bundlex/cnode.ex
Comment (mickel8) [inline lib/bundlex/cnode.ex:113]:
```suggestion
        %{app: app, native_name: native_name, caller: self(), link?: link?, cookie_env?: cookie_env?(app,native_name)}
```

File: lib/bundlex/cnode/server.ex
Comment (mickel8) [inline lib/bundlex/cnode/server.ex:16]:
```suggestion
    env = if opts.env?, do: [{'ERLANG_COOKIE', Atom.to_charlist(Node.get_cookie)} ],else: []
```

**Review by mickel8 — CHANGES_REQUESTED** [review]
Thanks a lot for the catch and PR! :slightly_smiling_face: 

Generally I think we can merge it. I suggested renaming a few variables. 
I will also wait for @mat-hek and @AndrzejPodo opinions.  
Before merging please use `mix format`.

File: lib/bundlex/cnode.ex
Comment (TypedLambda) [reply lib/bundlex/cnode.ex:104]: The code to resolve the config option is a bit complex,
i chose it to allow partial migrations in code using this.

the code above resolves the `cookie_env` option starting at the `app` config.
and if it's a map it loops up the native_name.
If it's not set, it gives the `:bundlex` config a try.

If per executable config is overkill, this can be dropped my (first commit should have this version).

I personally would never pass the cookie per argument, so i'll be setting `config :bundlex :cookie_env true` for all of my projects.
but this is a breaking change so it's not what i would suggest. 

---

Alternatively i would suggest to enforce an global or per-app config, and __fail__ if no option is set (to force users to choose an option).

It reduces complexity and improves security (by notifying explicitly about this option).
But it would force users to set the configuration upon update.

After thinking a bit about this...
For our commercial projects, and maybe even for public libraries I would maintain, I would probably chose a less configurable option and enforce a configuration.

**Comment by TypedLambda** [comment]
As stated in https://github.com/membraneframework/bundlex/pull/84#discussion_r568878866
I tried to make this configurable for every CNode binary, of every app.
To aid migration.

That might be overkill, only supporting one config per app might be enough.
then you would need to update all cnodes of an app at a time. 

I would like suggest to emit a warning to notify users to update the CNodes to get rid of the cookie in the process listing.

either whenever the cookie is passed via an argument 
or at least env_cookie config is unset (setting it to false disables the warning).

what do you think?

File: lib/bundlex/cnode.ex
Comment (mickel8) [reply lib/bundlex/cnode.ex]: Oh, now I see. Acctually bundlex hasn't been released as stable yet. It's still in initial development phase so making breaking changes shouldn't cost us much. For me switching entirely to passing cookie via env variable sounds even better but let's wait for @mat-hek with decision. He will be back on Monday so I think we can merge it then.

**Comment by mat-hek** [comment]
Hi @TypedLambda, I agree with @mickel8 that we should force users to switch to the new API (passing by env vars). The reasons are:
- it's a security vulnerability
- the required changes are easy to perform
- we're still in beta
- the solution with global configuration doesn't actually guarantee backwards compatibility. It would break if you configured it in an app A, depending on B, and at some point, B added a dependency to C, that only handles the old API.
- the solution with per-dependency configuration would require keeping track of whether the dependency has updated to the new API. Also, it doesn't help until all the dependencies do that
- configuration is only evaluated in top-level applications, not in intermediary libraries

**Comment by TypedLambda** [comment]
I Updated the change request again, removing the configuration code.
The environment variable is now always used for passing the cookie.

Due to the assert in the example code, CNodes based on that will fail due to the
lower number of arguments passed, which is also a great way to ensure the update is done to the C code.

History is left in the branch, feel free to squash it ;)

File: lib/bundlex/cnode.ex
Comment (mat-hek) [inline lib/bundlex/cnode.ex:113]: It seems to be unused now

**Review by mat-hek — CHANGES_REQUESTED** [review]

**Comment by TypedLambda** [comment]
Hold on, messed up the commit. contains old code i assumed i removed. fixing...

**Review by mat-hek — APPROVED** [review]

**Comment by mat-hek** [comment]
Thanks :) Will release today

**Pull request merged**
